Implement City Temperature Management API

vladyslav-tmf opened the pull request.

vladyslav-tmf committed:
Initialize Poetry & implement project settings

- Set up Poetry with fastapi dependency
- Implement settings.py with Pydantic settings
- Add .env.sample file
- Uncomment .idea/ in .gitignore

vladyslav-tmf committed:
Implement database configuration

- Set up SQLAlchemy engine and session management
- Add Base class for all database models
- Add database initialization functionality
- Configure database connection with SQLite

vladyslav-tmf committed:
Implement City & Temperature database models

vladyslav-tmf committed:
Implement City & Temperature Pydantic schemas

vladyslav-tmf committed:
Implement cities router with CRUD endpoints

vladyslav-tmf committed:
Implement weather service

- Create WeatherAPI service for fetching temperature data
- Add error handling for API requests

vladyslav-tmf committed:
Implement temperatures router

vladyslav-tmf committed:
Enhance API pagination and add base URL to page links

- Update city and temperature routers to include full base URL in pagination links
- Fix pagination schemas to use string-based page links
- Add timezone handling for temperature date validation
- Update .gitignore to exclude database files
- Create main FastAPI application with router configuration

vladyslav-tmf committed:
Create project documentation

- Add detailed project description
- Add installation and running instructions
- Add project structure overview

mateacademy-ai-mentor commented on app/config/settings.py:
```diff
@@ -0,0 +1,23 @@
+
+class Settings(BaseSettings):
+
+    weather_api_key: str = os.getenv("WEATHER_API_KEY")
```
Using `os.getenv` directly for environment variables is not necessary when using Pydantic's `BaseSettings`. You can define these variables directly in the class, and Pydantic will automatically load them from the environment. Consider changing `weather_api_key: str = os.getenv("WEATHER_API_KEY")` to `weather_api_key: str` and let Pydantic handle the environment variable loading.

mateacademy-ai-mentor commented on app/config/settings.py:
```diff
@@ -0,0 +1,23 @@
+class Settings(BaseSettings):
+
+    weather_api_key: str = os.getenv("WEATHER_API_KEY")
+    weather_api_base_url: str = os.getenv(
+        "WEATHER_API_BASE_URL", "http://api.weatherapi.com/v1"
+    )
```
Similarly, for `weather_api_base_url`, you can define it directly in the class without using `os.getenv`. Pydantic will use the default value if the environment variable is not set. Change `weather_api_base_url: str = os.getenv("WEATHER_API_BASE_URL", "http://api.weatherapi.com/v1")` to `weather_api_base_url: str = "http://api.weatherapi.com/v1"`.

mateacademy-ai-mentor commented on app/database/models/city.py:
```diff
@@ -0,0 +1,18 @@
+    )
+    additional_info: Mapped[str | None] = mapped_column(String(255), nullable=True)
+
+    temperatures: Mapped[list["Temperature"]] = relationship(  # noqa: F821
+        back_populates="city", cascade="all, delete-orphan"
+    )
```
The forward reference to the `Temperature` model in the `temperatures` relationship might cause issues if the `Temperature` model is not defined in the same module or imported before this class is used. Ensure that the `Temperature` model is correctly imported or defined to avoid runtime errors.

mateacademy-ai-mentor commented on app/database/models/temperature.py:
```diff
@@ -0,0 +1,21 @@
+    )
+    temperature: Mapped[float] = mapped_column(nullable=False)
+
+    city: Mapped["City"] = relationship(back_populates="temperatures")  # noqa: F821
```
The forward reference to the `City` model in the `city` relationship might cause issues if the `City` model is not defined in the same module or imported before this class is used. Ensure that the `City` model is correctly imported or defined to avoid runtime errors.

mateacademy-ai-mentor commented on app/routers/city.py:
```diff
@@ -0,0 +1,118 @@
+) -> PaginatedCityResponseSchema:
+    cities = db.query(City).offset((page - 1) * per_page).limit(per_page).all()
+
+    if not cities:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="No cities found",
+        )
```
Raising a 404 error when no cities are found might not be necessary for a paginated list endpoint. Consider returning an empty list with a 200 status code instead, as this is a common practice for paginated endpoints.

mateacademy-ai-mentor commented on app/routers/city.py:
```diff
@@ -0,0 +1,118 @@
+            detail="No cities found",
+        )
+
+    city_list = [CityResponseSchema.model_validate(city) for city in cities]
```
Ensure that `CityResponseSchema.model_validate(city)` correctly validates and serializes the city object. If there are any issues with the schema, it might lead to runtime errors.

mateacademy-ai-mentor commented on app/routers/temperature.py:
```diff
@@ -0,0 +1,106 @@
+async def update_temperatures(db: Session = Depends(get_db)) -> dict[str, str]:
+    cities = db.query(City).all()
+
+    if not cities:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="No cities found in the database",
+        )
```
Raising a 404 error when no cities are found might not be necessary for the update endpoint. Consider handling this case more gracefully, perhaps by logging a message or returning a different status code.

mateacademy-ai-mentor commented on app/routers/temperature.py:
```diff
@@ -0,0 +1,106 @@
+        .all()
+    )
+
+    if not temperatures and page > 1:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="No temperatures found",
+        )
```
Raising a 404 error when no temperatures are found on a specific page might not be necessary for a paginated list endpoint. Consider returning an empty list with a 200 status code instead, as this is a common practice for paginated endpoints.

mateacademy-ai-mentor commented on app/routers/temperature.py:
```diff
@@ -0,0 +1,106 @@
+
+    return PaginatedTemperatureResponseSchema(
+        temperatures=[
+            TemperatureResponseSchema.model_validate(temp) for temp in temperatures
```
Ensure that `TemperatureResponseSchema.model_validate(temp)` correctly validates and serializes the temperature object. If there are any issues with the schema, it might lead to runtime errors.

mateacademy-ai-mentor reviewed (approved):
****

Your code is well-structured and meets the core requirements of the task. 🎉 You've done a great job implementing the City and Temperature APIs with FastAPI, and your use of SQLAlchemy and Pydantic is on point. While there are some minor improvements suggested, such as handling empty paginated lists more gracefully and optimizing environment variable handling with Pydantic, these do not critically impact the functionality of your code. Keep up the great work, and consider the feedback for future enhancements! 😊

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

AnyoneClown commented on app/routers/temperature.py:
```diff
@@ -0,0 +1,106 @@
+)
+from app.services.weather import get_current_temperatures
+
+temperature_router = APIRouter(prefix="/temperatures", tags=["temperatures"])
```
Usually, the router is called router, and then when importing, use the `import router as some_router` naming convention. Refactor this

AnyoneClown commented on app/routers/temperature.py:
```diff
@@ -0,0 +1,106 @@
+    city_names = [city.name for city in cities]
+
+    async with aiohttp.ClientSession() as session:
+        temperatures = await get_current_temperatures(city_names, session)
+
```
You are raising errors in this method, but you are not trying to catch them here. Use `try/except` logic

vladyslav-tmf committed:
Refactor routers and settings configuration

- Update router imports variables in main.py and routers
- Update settings to use Pydantic Field and remove direct os.getenv calls
- Add error handling for temperature updates in temperature router
